### DISCUSSION TITLE
Seemingly bug with custom enums in Diesel SQLite?

### DISCUSSION THREAD

**Discussion by ksaadDE** [diesel-rs/diesel#3742] [Q&A] [answered]
### Versions

- **Rust:**  rustc 1.71.0
- **Diesel:** `diesel = {version="2.1.0", features = ["sqlite", "returning_clauses_for_sqlite_3_35", "chrono"] }` 
- **Database:** SQLite
- **Operating System:**  Ubuntu

## Problem Description
```the trait `diesel::Expression` is not implemented for 'TaskStatus'```

### What are you trying to accomplish?
I try to get the enum val stored in the DB...

### What is the expected output?
TaskStatus is converted to a string to be stored in the DB

### What is the actual output?
nothing just errors by diesel

### Are you seeing any additional errors?
```rust
error[E0277]: the trait bound `TaskStatus: diesel::Queryable<diesel::sql_types::Text, Sqlite>` is not satisfied
  --> src/models.rs:59:17
   |
59 |     pub status: TaskStatus,
   |                 ^^^^^^^^^^ the trait `diesel::Queryable<diesel::sql_types::Text, Sqlite>` is not implemented for `TaskStatus`
   |
   = help: the following other types implement trait `diesel::Queryable<ST, DB>`:
             <(T0, T1) as diesel::Queryable<(ST0, ST1), __DB>>
             <(T0, T1, T2) as diesel::Queryable<(ST0, ST1, ST2), __DB>>
             <(T0, T1, T2, T3) as diesel::Queryable<(ST0, ST1, ST2, ST3), __DB>>
             <(T0, T1, T2, T3, T4) as diesel::Queryable<(ST0, ST1, ST2, ST3, ST4), __DB>>
             <(T0, T1, T2, T3, T4, T5) as diesel::Queryable<(ST0, ST1, ST2, ST3, ST4, ST5), __DB>>
             <(T0, T1, T2, T3, T4, T5, T6) as diesel::Queryable<(ST0, ST1, ST2, ST3, ST4, ST5, ST6), __DB>>
             <(T0, T1, T2, T3, T4, T5, T6, T7) as diesel::Queryable<(ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7), __DB>>
             <(T0, T1, T2, T3, T4, T5, T6, T7, T8) as diesel::Queryable<(ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7, ST8), __DB>>
           and 80 others
   = note: required for `TaskStatus` to implement `FromSqlRow<diesel::sql_types::Text, Sqlite>`
   = help: see issue #48214

error[E0277]: the trait bound `TaskStatus: diesel::Expression` is not satisfied
  --> src/models.rs:71:9
   |
71 |     pub status: &'a TaskStatus,
   |         ^^^^^^ the trait `diesel::Expression` is not implemented for `TaskStatus`
   |
   = help: the following other types implement trait `diesel::Expression`:
             &'a T
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
             (T0, T1, T2, T3, T4, T5)
             (T0, T1, T2, T3, T4, T5, T6)
             (T0, T1, T2, T3, T4, T5, T6, T7)
           and 86 others
   = note: required for `&'a TaskStatus` to implement `diesel::Expression`
   = note: required for `&'a TaskStatus` to implement `diesel::expression::AsExpression<diesel::sql_types::Text>`

error[E0277]: the trait bound `TaskStatus: diesel::Expression` is not satisfied
  --> src/models.rs:71:9
   |
71 |     pub status: &'a TaskStatus,
   |         ^^^^^^ the trait `diesel::Expression` is not implemented for `TaskStatus`
   |
   = help: the following other types implement trait `diesel::Expression`:
             &'a T
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
             (T0, T1, T2, T3, T4, T5)
             (T0, T1, T2, T3, T4, T5, T6)
             (T0, T1, T2, T3, T4, T5, T6, T7)
           and 86 others
   = note: required for `&'a TaskStatus` to implement `diesel::Expression`
   = note: required for `&'insert &'a TaskStatus` to implement `diesel::expression::AsExpression<diesel::sql_types::Text>`

```

### Steps to reproduce

```rust
use super::schema::tasks;
use diesel::prelude::*;
use diesel::*;
use diesel::deserialize::{self, FromSql};

//use diesel::sql_types::Nullable;

use chrono::NaiveDateTime;

//use diesel::sqlite::Sqlite;
use std::io::Write;
use diesel::serialize::{self, IsNull, Output, ToSql};
use diesel::deserialize::{self, FromSql};

use diesel::sqlite::Sqlite;

#[derive(Debug, PartialEq, AsExpression)]
pub enum TaskStatus {
    New,
    OnHold,
    InProcess,
    Done
}

#[derive(Queryable, Selectable)]
#[diesel(table_name = crate::schema::tasks)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Task {
    pub id: i32,
    pub name: String,
    pub description: String,
    pub status: TaskStatus,
    pub isUrgent: bool,
    pub isImportant: bool,
    pub deadline: NaiveDateTime,
    pub deleted: bool
}

#[derive(Insertable)]
#[diesel(table_name = tasks)]
pub struct NewTask<'a> {
    pub name: &'a String,
    pub description: &'a String,
    pub status: &'a TaskStatus,
    pub isUrgent: &'a bool,
    pub isImportant: &'a bool,
    pub deadline: &'a NaiveDateTime,
}



#[cfg(feature = "diesel_sqlite")]
impl<Sqlite> ToSql<TaskStatus, Sqlite> for TaskStatus {
    fn to_sql<W: Write>(&self, out: &mut Output<W, '_, Sqlite>) -> serialize::Result {
        match *self {
            TaskStatus::New => out.write_all(b"New")?,
            TaskStatus::OnHold => out.write_all(b"OnHold")?,
            TaskStatus::InProcess => out.write_all(b"InProcess")?,
            TaskStatus::Done => out.write_all(b"Done")?,
        }
        Ok(IsNull::No)
    }
}


#[cfg(feature = "diesel_sqlite")]
impl FromSql<TaskStatus, Sqlite> for TaskStatus {
    fn from_sql(bytes: Option<&[u8]>) -> deserialize::Result<Self> {
        match bytes {
            b"New" => Ok(TaskStatus::New),
            b"OnHold" => Ok(TaskStatus::OnHold),
            b"InProcess" => Ok(TaskStatus::OnHold),
            b"Done" => Ok(TaskStatus::OnHold),
            _ => Err("Unrecognized enum variant".into()),
        }
    }
}
```
 
 ```
 diesel::table! {
    tasks (id) {
        id -> Integer,
        name -> Text,
        description -> Text,
        isUrgent -> Bool,
        isImportant -> Bool,
        deadline -> Timestamp,
        status -> Text,
        deleted -> Bool,
    }
}

 ```
 
I'm new to Rust. Don't expect me to know anything or any of your terminology. I just want to store the enum value in DB. I searched the whole documentation, the internet and I looked at the custom_types.rs which doesn't help.  I tried different variants... Text fields... Timestamp... copy pasta... literally all that is possible...Any ideas? 

I have read that you might have introduced some bugs in the new version. Although the code here is older than just a few days. Still it might be possible that this is some kind of bug. For some reason I had the same issues with Timestamps!

Additionally better examples and updated docs would be very nice in the future. Like for a greenhorn it is far too much effort to get used to that kind of basic stuff within Diesel. 

**/E: I solved my problem, see here:** 
- [Enum <-string-> DB Documentation by ksaadDE](https://github.com/ksaadDE/rustgotchas/blob/main/How_to_convert_a_Enum_to_DB_row_string_value_using_Diesel.md)

**Comment by weiznich** [comment]
I've converted this to a support question, as that's not a bug in diesel. Generally speaking: Please do not open bug reports for support questions. The next case will be handled as "bug report", which means it's just closed as: That's not a bug without any support help.

It's just a not working implementation on your side.

There are quite a view issues in your code. First of all your error message is missing an important part:

```
error: At least one `sql_type` is needed for deriving `AsExpression` on a structure.
  --> src/main.rs:31:28
   |
31 | #[derive(Debug, PartialEq, AsExpression)]
   |                            ^^^^^^^^^^^^
   |
   = note: this error originates in the derive macro `AsExpression` (in Nightly builds, run with -Z macro-backtrace for more info)
```

which points out that your `#[derve(AsExpression)]` attribute is incomplete. 

The relevant documentation for implementing custom type support via diesel itself is [here (`FromSql`)](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html#examples) and [here (`ToSql`)](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html#examples). These examples (as all rustdoc examples) are tested on our CI so it's highly unlikely that we accidentally broke them.

To list all the important differences between your code those both examples:

* Your enum is missing a `#[derive(FromSqlRow)]` as required for implementing `FromSql`
* Your enum is missing a `#[diesel(sql_type = diesel::sql_types::Text)]` attribute as required for deriving `#[derive(AsExpression)]` as shown in the documentation + requested by the error message
* Your `FromSql` implementation is wrong. It needs to use an sql type (matching your column type) as generic parameter. (It needs to be `impl FromSql<Text, Sqlite> for TaskStatus {…}`
* Your `FromSql` implementation indicates that you used old (diesel 1.x) postgres code as example. Follow the function signatures and implementation as shown in the linked `FromSql` example.
* Your `ToSql` implementation is wrong. It also need to use a matching sql type as generic parameter.
* Your `ToSql` implementation should not use `Sqlite` as generic type. Instead import the `diesel::sqlite::Sqlite` backend type.
* Your `ToSql` implementation indicates that you followed old (diesel 1.x) postgres code as example. Follow the function signatures of the linked `ToSql` example.




> Additionally better examples and updated docs would be very nice in the future. Like for a greenhorn it is far too much effort to get used to that kind of basic stuff within Diesel.

It's always easy to ask for "better documentation" and examples in a generic way. It's hard to make specific improvements, so let me ask that way: What exactly would you like to improve? Please propose a concrete improvement.

**Reply by ksaadDE** [reply]
@weiznich  Thank you very much for your reply and help.

## V1 reply
> attribute as required for deriving #[derive(AsExpression)] as shown in the documentation + requested by the error message

For example, the docs do not show any example how to implement this or a tiny hint that this would be needed. Yes, there is a codebox that shows that derive but it does not show that it is needed! So how should we newbies know what needs to be done to get it working?

> `diesel::expression::AsExpression<diesel::sql_types::Text>`

The last line of the error message does not make me think of adding `#[diesel(sql_type = diesel::sql_types::Text)]` because of the errorr message that `#[derive(AsExpression)]` throws.


> `import the diesel::sqlite::Sqlite`

I guess it was already imported (js the imports were mixed) 

## V2 - new try
I updated the code according to your suggestions!  It still does not compile or work. I'm banging my head against the wall for hours.


**New code:**
```rust
use std::string::ToString;
use strum_macros::Display;


use super::schema::tasks;

use std::io::Write;

use diesel::prelude::*;
use diesel::*;
use diesel::sqlite::Sqlite; //use diesel::sqlite::Sqlite;
use diesel::deserialize::{self, FromSql};
use diesel::serialize::{self, IsNull, Output, ToSql};
use diesel::sql_types::Text;
use diesel::sqlite::SqliteValue;
//use diesel::sql_types::Nullable;
use chrono::NaiveDateTime;


#[repr(i32)]
#[derive(SqlType, Debug, PartialEq, FromSqlRow, AsExpression, Display)]
#[diesel(sql_type = diesel::sql_types::Text)]
pub enum TaskStatus {
    New,
    OnHold,
    InProcess,
    Done
}

#[derive(Queryable, Selectable)]
#[diesel(table_name = crate::schema::tasks)]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
pub struct Task {
    pub id: i32,
    pub name: String,
    pub description: String,
    pub status: TaskStatus,
    pub isUrgent: bool,
    pub isImportant: bool,
    pub deadline: NaiveDateTime,
    pub deleted: bool
}

#[derive(Insertable)]
#[diesel(table_name = tasks)]
pub struct NewTask<'a> {
    pub name: &'a String,
    pub description: &'a String,
    pub status: &'a TaskStatus,
    pub isUrgent: &'a bool,
    pub isImportant: &'a bool,
    pub deadline: &'a NaiveDateTime,
}

#[cfg(feature = "diesel_sqlite")]
impl FromSql<sql_types::Text, diesel::sqlite::Sqlite> for TaskStatus {
    fn from_sql(value: SqliteValue<'_, '_, '_>) -> deserialize::Result<Self> {
        let text = value.read_text();
        Ok(text)
    }
}

impl ToSql<sql_types::Text, diesel::sqlite::Sqlite> for TaskStatus
where
    str: ToSql<sql_types::Text, diesel::sqlite::Sqlite>,
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, diesel::sqlite::Sqlite>) -> serialize::Result {
        let v = *self as str;
        <str as ToSql<sql_types::Text, diesel::sqlite::Sqlite>>::to_sql(&v, &mut out.reborrow())
    }
}

```

**Compiler output:**
```rust
error[E0277]: the trait bound `TaskStatus: FromSql<diesel::sql_types::Text, Sqlite>` is not satisfied
  --> src/models.rs:37:17
   |
37 |     pub status: TaskStatus,
   |                 ^^^^^^^^^^ the trait `FromSql<diesel::sql_types::Text, Sqlite>` is not implemented for `TaskStatus`
   |
   = help: the following other types implement trait `FromSql<A, DB>`:
             <*const [u8] as FromSql<diesel::sql_types::Binary, Sqlite>>
             <*const str as FromSql<diesel::sql_types::Text, Sqlite>>
             <Cow<'a, T> as FromSql<ST, DB>>
             <MigrationVersion<'a> as FromSql<diesel::sql_types::Text, DB>>
             <NaiveDate as FromSql<diesel::sql_types::Date, Sqlite>>
             <NaiveDateTime as FromSql<Timestamptz, Sqlite>>
             <NaiveDateTime as FromSql<diesel::sql_types::Timestamp, Sqlite>>
             <NaiveTime as FromSql<diesel::sql_types::Time, Sqlite>>
           and 15 others
note: required for `TaskStatus` to implement `diesel::Queryable<diesel::sql_types::Text, Sqlite>`
  --> src/models.rs:21:37
   |
21 | #[derive(SqlType, Debug, PartialEq, FromSqlRow, AsExpression, Display)]
   |                                     ^^^^^^^^^^ unsatisfied trait bound introduced in this `derive` macro
22 | #[diesel(sql_type = diesel::sql_types::Text)]
23 | pub enum TaskStatus {
   |          ^^^^^^^^^^
   = note: required for `TaskStatus` to implement `diesel::deserialize::FromSqlRow<diesel::sql_types::Text, Sqlite>`
   = help: see issue #48214
   = note: this error originates in the derive macro `FromSqlRow` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```

**Reply by ksaadDE** [reply]
Seemingly, I am not the only one having that type of issues. I just have it for SQLite not for Pg

- https://www.reddit.com/r/rust/comments/14jg3km/enum_does_not_work_with_diesel/
- https://users.rust-lang.org/t/diesel-first-steps-and-hours-of-frustration/93318

This one got me closer to a working solution
- https://stackoverflow.com/questions/72273087/error-creating-custom-type-wrappers-in-diesel/72282900#72282900

It lead to this:
```rust
impl ToSql<Text, Sqlite> for TaskStatus {
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Sqlite>) -> serialize::Result {
        <String as ToSql<Text, Sqlite>>::to_sql(&self.to_string(), &mut out.reborrow())
    }
}

impl FromSql<Text, Sqlite> for TaskStatus {
    fn from_sql(bytes: <Sqlite as Backend>::RawValue<'_>) -> deserialize::Result<Self> {
        Ok(<String as FromSql<Text, Sqlite>>::from_sql(bytes).map(|s| TaskStatus::from_str(&s).unwrap())?)
    }
}
```

But throws
```rust
error[E0599]: the method `reborrow` exists for mutable reference `&mut Output<'b, '_, Sqlite>`, but its trait bounds were not satisfied
  --> src/models.rs:67:77
   |
67 |         <String as ToSql<Text, Sqlite>>::to_sql(&self.to_string(), &mut out.reborrow())
   |                                                                             ^^^^^^^^ method cannot be called on `&mut Output<'b, '_, Sqlite>` due to unsatisfied trait bounds
   |
  ::: /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/diesel-2.1.0/src/sqlite/backend.rs:10:1
   |
10 | pub struct Sqlite;
   | ----------------- doesn't satisfy `<_ as Backend>::BindCollector<'c> = RawBytesBindCollector<Sqlite>`
   |
   = note: the following trait bounds were not satisfied:
           `<Sqlite as Backend>::BindCollector<'c> = RawBytesBindCollector<Sqlite>`

Some errors have detailed explanations: E0252, E0599.
For more information about an error, try `rustc --explain E0252`.
```

Tbh no clue how to fix it!

**Reply by weiznich** [reply]
> For example, the docs do not show any example how to implement this or a tiny hint that this would be needed. Yes, there is a codebox that shows that derive but it does not show that it is needed! So how should we newbies know what needs to be done to get it working?

I'm not sure at which documentation you are looking here because thats really in there. Let me link and highlight the relevant parts:

The [`ToSql`](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html) states that deriving `AsExpression` is required.

> Any types which implement this trait should also #[derive(AsExpression)].

This links [the documentation of that derive](https://docs.diesel.rs/2.1.x/diesel/expression/derive.AsExpression.html#required-container-attributes), which contains a section on required container attributes, which states that `#[diesel(sql_type = SqlType)]` must exist. 

>`#[diesel(sql_type = SqlType)]`, to specify the sql type of the generated implementations. If the attribute exists multiple times impls for each sql type is generated.

The only thing that's arguably not in there is which type exactly needs to be used there and that depends on your needs. Additionally the "code box" on the `ToSql` documentation page shows a complete self contained example, so I'm unsure which part exactly you consider to be missing or why you believe that you can just skip parts of an example and the complain why its not working.


The error message in your first post just indicates that the `diesel_sqlite` feature is not enabled.

For the error message in the second post let me quote again [from the documentation](https://docs.diesel.rs/2.1.x/diesel/serialize/trait.ToSql.html#examples):

> Using temporary values as part of the ToSql implementation requires additional work.

> Backends using RawBytesBindCollector as BindCollector copy the serialized values as part of Write implementation. This includes the Mysql and the Pg backend provided by diesel. This means existing ToSql implementations can be used even with temporary values. For these it is required to call Output::reborrow to shorten the lifetime of the Output type correspondingly.

> For any other backend the Output::set_value method provides a way to set the output value directly. Checkout the documentation of the corresponding BindCollector::Buffer type for provided From<T> implementations for a list of accepted types. For the Sqlite backend see SqliteBindValue.

It also contains examples for both case. And just in case you miss the relevant documentation again: The `From` impls are listed [here](https://docs.diesel.rs/2.1.x/diesel/sqlite/struct.SqliteBindValue.html#trait-implementations)

Honestly I have the impression the the issue here is not missing documentation, but more persons that just skim the documentation and then conclude that it must work this or that way. I do not see how we could solve this by adding even more documentation. It seems to be more an combination of trying the better highlight the important stuff and people that are just not familiar with reading rust api documentation. To be clear, both is not your fault. Its something we are actively looking for suggestions on how to improve the current situation.

**Reply by ksaadDE** [reply]
Thank you very much for your reply.

Is it even possible to get my code running? Yet I have not found any example of that on the inet. I spent about 16 hours only on that simple task. Keep in mind in e.g. Doctrine + PHP Symfony that is done in like 20 minutes (prob less). I do not see why your ORM should be that special that we can not do it in less time.

> Additionally the "code box" on the ToSql documentation page shows a complete self contained example,

For my case? Or just another generalized Pg example, which is not helpful rn? I copy & pasted from your docs and adapted it. After that did not work for yet unknown reasons (prob that Buffer thingy) I searched like two hours on the inet (Stackoverflow etc) it still does not work. If someone of you got it working, lmk.

> more persons that just skim the documentation and then conclude that it must work this or that way.

Obviously nobody really reads 300+ pages documentation. Most people read what they need to do right now and scan only for the code examples. After that they copy paste and adapt it to their needs, until it works. Then they do their next task. Like every normal human.

You perfectly highlight that point in your next statement here:

>  highlight the important stuff and people that are just not familiar with reading rust api documentation [...] Its something we are actively looking for suggestions on how to improve the current situation.

That problem is increased by using extra terminology nobody can connect to, especially greenhorns have that issue, like you have seen here. 

How should I know that your Buffer thingy is somewhat connected to the simple task of implementing your ToSQL / FromSQL stuff? It does not make sense to write a whole paragraph about a thing nobody without internal knowledge can connect to. 

Like in this case, I just want the string row val in my DB.. I do not see a point for digging deeper than that in the inner workings of your ORM. As that is a overhead that is absolutely not needed for such a default task. 

Please understand me the right way: It IS nice to learn over time but if I start with your ORM I do not want to dig that deep, as it just frustrates me (probably everyone else out there too) and makes things much harder to accomplish than needed!

> The From impls are listed [here](https://docs.diesel.rs/2.1.x/diesel/sqlite/struct.SqliteBindValue.html#trait-implementations)

What do I have to look for? I see no code there (= no examples). I just see raw code documentation style. That actually helps if you already have a skeleton or basic idea to just look up the types. But they are not helpful in this case as the whole approach is still missing. 

How can I use that in my case and where can I see how to implement that Buffer thingy? How much time do I have to plan for that? 


A little help would be great, a little code example for that specific case. Others of the community would profit from that too, without having the crazy effort I had to invest. Like I have shown many people have that issue.

**Reply by ksaadDE** [reply]
Kindly check the Cargo.toml to identify if `diesel_sqlite` is not loaded

```
[package]
name = "name"
version = "0.1.0"
edition = "2021"

[dependencies]
argon2 = "0.5.0"
cargo-watch = "8.4.0"
chrono = "0.4.26"
diesel = {version="2.1.0", features = ["sqlite", "returning_clauses_for_sqlite_3_35", "chrono"] }
dotenvy = "0.15.7"
hex = "0.4.3"
hex-literal = "0.4.1"
password-hash = "0.5.0"
pbkdf2 = { version = "0.12.2", features = ["simple"] }
rand_core = "0.6.4"
rocket = { version = "0.5.0-rc.2", features = ["json"] }
scrypt = "0.11.0"
sha2 = "0.10.7"
strum = "0.25.0"
strum_macros = "0.25.2"
watch = "0.2.3"

[dependencies.rocket_dyn_templates]
version = "=0.1.0-rc.3"
features = ["handlebars", "tera"]

```

Again here, the documentation and code examples are not really helpful. Too much individual pieces, conflicting versions and dependencies that are either too much or not mentioned at all. At least that is my impression and I think many others have that impression too (compare StackOverflow or the issues / Q&A tab).

I get that running such a complicated project is really time intensive and not easily done. The documentation is a lot of extra work and effort. Especially switching from the mind "in the project" to the "think as greenhorn"-mind is really hard. But that must be done!

Think like the absolute noob that wants to get simple things done. They might have a little bit of experience in Python or any other language out there but they need to get SLOWLY fluent with your ORM. So that should be the starting point and nothing else.

Also, it is really bad if you mix up ten different ways of doing a thing in one file/example. For example, the `custom_types.rs` (I am right?) does that. There is so much bunched together that nobody can really differentiate what is part of what. 

I do not want to bash you. I have the highest respect for your work. I could never do that in my human time. Like you have done it. But that appreciation does not prevent me from pointing out these critical issues with the docs and the example code. 

I would like to provide the fix for my code as an example to implement it, so others can copy & paste it, to learn on the fly. Sadly, I can not solve it right now. I still need figure out the right solution. I highly appreciate your help in advance. 🙏

**Comment by weiznich** [comment]
> For my case? Or just another generalized Pg example, which is not helpful rn? I copy & pasted from your docs and adapted it. After that did not work for yet unknown reasons (prob that Buffer thingy) I searched like two hours on the inet (Stackoverflow etc) it still does not work. If someone of you got it working, lmk.

I'm not sure how often I should repeat that… The examples in the linked documentation are **not** postgres specific. they work with all backends unless that it's explicitly spelled out in the text that a certain example is for a specific backend only. That means for the `ToSql` documentation: The first example works for all backend, the second example is postgres only (although it would work with minimal modifications using the `Mysql` backend as well) and the third example is sqlite only. The first and third example covers your use-case, depending on how you exactly want to implement the trait. I really cannot see how providing another example would be helpful for you here, as it would be basically the same example as provided by the documentation.

And just to prevent further discussion on that point: If you believe the documentation is not right in that place, I would like to see a bit more evidence than: I've copied and *modified* the code and now it doesn't work. It's kind of expected that code breaks if you modify it without understanding what it does.

> Obviously nobody really reads 300+ pages documentation. Most people read what they need to do right now and scan only for the code examples. After that they copy paste and adapt it to their needs, until it works. Then they do their next task. Like every normal human.

Well maybe you shouldn't complain about missing documentation then if you don't brother reading the documentation at all. It's right that most people don't read the whole documentation at one, but most people go back to the documentation and start reading more of it as soon as something does not work as expected. That seems not to be the case for you, maybe consider doing that the next time?

> What do I have to look for? I see no code there (= no examples). I just see raw code documentation style. That actually helps if you already have a skeleton or basic idea to just look up the types. But they are not helpful in this case as the whole approach is still missing.

If you are not familiar with the `From` trait, you should consider going back to rust book and start again with the basics of the language. It's something that's really considered to be part of the language skills here.

> Kindly check the Cargo.toml to identify if diesel_sqlite is not loaded

It's impossible to check that for sure based on the provided code, as I have no way of knowing whether that's your complete `Cargo.toml` or not. **If** it's the complete `Cargo.toml` then there is no feature with the name `diesel_sqlite` defined. In that case I would suggest that you go back to the rust book and read about cargo features.

> I get that running such a complicated project is really time intensive and not easily done. The documentation is a lot of extra work and effort. Especially switching from the mind "in the project" to the "think as greenhorn"-mind is really hard. But that must be done!

Let me make that as clear as possible: I do not need to do anything here. The code is provided under the `MIT` license with states, which explicitly states the code is provided "as it is". You and no other user have a right to claim something from me. That includes writing documentation and also providing support. You are free to use this code and I'm normally happy with providing at least some kind of support, although my motivation to further answer in this thread is mostly gone. If you need more than the support/documenation level provided by default we can certainly talk about a consulting contract, otherwise you are on your own.

>  For example, the custom_types.rs (I am right?) does that. There is so much bunched together that nobody can really differentiate what is part of what.

`custom_types.rs` does something slightly different. It implements a fully custom user defined sql side enum type for postgresql. It's a different use case.

**Reply by ksaadDE** [reply, marked as answer]
Very interesting...,thanks for your reply. 

>  there is no feature with the name diesel_sqlite defined

Alright! Lets test that. 

We add a fully new entry for `diesel_sqlite` into the features array of the diesel dependency in `cargo.toml`.

Lets see what the new error message says:
```the package `<packname>` depends on `diesel`, with features: `diesel_sqlite` but `diesel` does not have these features```

If you would have checked the `cargo.toml` you would have seen that the `sqlite` diesel feature was already defined. It is actually the first entry of the Diesel dependency. So you are wrong with that point.

> with the From trait

At that point I was curious if you are evening understanding what you are talking about. Because it is not about Rust's `From` Trait. 

The described problem is about the implementation of the Diesel ORM's converter methods "`to_sql`" and "`from_sql`". Those need to be implemented, so that the ORM can convert between Enum and String, to and from the DB (string <-> enum). 

**The following works:**
```rust
#[repr(i32)]
#[derive(SqlType, Debug, PartialEq, FromSqlRow, AsExpression, Display, Copy, Clone, EnumString)]
#[diesel(sql_type = diesel::sql_types::Text)]
pub enum TaskStatus {
    New,
    OnHold,
    InProgress,
    Done
}

impl ToSql<Text, diesel::sqlite::Sqlite> for TaskStatus
where
    str: ToSql<Text, diesel::sqlite::Sqlite>,
{
    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, diesel::sqlite::Sqlite>) -> serialize::Result {
        out.set_value(self.to_string());
        Ok(IsNull::No)
    }
}

impl FromSql<Text, Sqlite> for TaskStatus {
    fn from_sql(bytes: <Sqlite as Backend>::RawValue<'_>) -> deserialize::Result<Self> {
        Ok(<String as FromSql<Text, Sqlite>>::from_sql(bytes).map(|s| TaskStatus::from_str(&s).unwrap())?)
    }
}
```

Proof that it works:
![working_show_tasks](https://github.com/diesel-rs/diesel/assets/37943746/2498bf80-b731-45f2-b21f-86ae8febc066)
![working_show_tasks_2](https://github.com/diesel-rs/diesel/assets/37943746/3d517f8e-0e26-4c98-98a8-8d8c4500ef4c)

Interestingly your last point was about the `BindCollector::Buffer`. Like I previously stated, the documentation and your replies are missing any central description or plan how to implement it in a way that it is working. More precisely the Buffer stuff, it is not even needed in that context. 

The links you send in your last message do not work, they do not point to code examples. Those just have the high-level technical documentation for the data types. Either you referred to the wrong doc page or you got confused. Lets stop at that point here.

The last example of the ToSql and FromSQL gave me a hint what I could do, you indirectly referred to that, without explaining anything. 

> we can certainly talk about a consulting contract, otherwise you are on your own

Why should I want to hire a consultant that keeps insulting the potential customer publicly?  Despite it is a private, solely educational project, which I am developing in my spare time... 

>  how often I should repeat 

After spending various hours on your examples, I asked for help, that alone feels embarrassing enough.
Were you seriously about to help or did you just try to exploit the missing liability in consulting contracts which you strongly pointed out before? 

Thanks for a wise reply in advance.

**Reply by ksaadDE** [reply]
I put my experiences in a small write up, in case any other people encounter such issues. 
[How_to_convert_a_Enum_to_DB_row_string_value_using_Diesel.md](https://github.com/ksaadDE/rustgotchas/blob/main/How_to_convert_a_Enum_to_DB_row_string_value_using_Diesel.md)

**Reply by weiznich** [reply]
It's hard to not write a sarcastic response here, nevertheless let me try that to address at least some of your wild claims…

As a general advice if you try to interact again with the project: Try to read the linked documentation and try to consider that you might be wrong about basics if you are new to a language or a project.

> We add a fully new entry for diesel_sqlite into the features array of the diesel dependency in cargo.toml.
> 
> Lets see what the new error message says:
the package `<packname>` depends on `diesel`, with features: `diesel_sqlite` but `diesel` does not have these features
>
> If you would have checked the cargo.toml you would have seen that the sqlite diesel feature was already defined. It is actually the first entry of the Diesel dependency. So you are wrong with that point.

This perfectly demonstrates what I've written above. If you use `#[cfg(feature = "diesel_sqlite")]` in your code it does **not** refer to the `sqlite` feature of the diesel crate, but to an feature named `diesel_sqlite` in your crate. The `sqlite` feature of diesel is completely unrelated to this. The original error message indicates that this feature does not exist or is not enabled, which results in the `FromSql` implementation not being compiled at all. This demonstrates a fundamental misunderstanding in how rust features work. It's not related to diesel at all.


> At that point I was curious if you are evening understanding what you are talking about. Because it is not about Rust's From Trait.
> 
> The described problem is about the implementation of the Diesel ORM's converter methods "to_sql" and "from_sql". Those need to be implemented, so that the ORM can convert between Enum and String, to and from the DB (string <-> enum).

Well at that point I'm not sure if you even took the time to read all I've wrote above, including the relevant documentation part. Nevertheless again: 

> For any other backend the [Output::set_value](https://docs.diesel.rs/2.1.x/diesel/serialize/struct.Output.html#method.set_value) method provides a way to set the output value directly. Checkout the documentation of the corresponding BindCollector::Buffer type for provided From<T> implementations for a list of accepted types. For the Sqlite backend see SqliteBindValue.

It literally states that you need to call `out.set_value()` for `Sqlite` + links the relevant function signature. This function signature includes a `Into<>` trait bound, which in turn implies that a `From` impl needs to exist as also mentioned by the documentation.  (The `Into`/`From` duality is some basic rust knowledge). The linked `From` implementations are relevant to know which types can be passed to `set_value. 
On a more general note: People coming from dynamically typed languages normally miss that type information of function/trait/… signatures carry a lot information on their own. It's common in rust to not repeat that information in prosa, so try learn how to read that documentation or you will have a hard time understanding any rust documentation.

> I put my experiences in a small write up, in case any other people encounter such issues.
> [How_to_convert_a_Enum_to_DB_row_string_value_using_Diesel.md](https://github.com/ksaadDE/rustgotchas/blob/main/How_to_convert_a_Enum_to_DB_row_string_value_using_Diesel.md)

Thanks for providing that document. For any future reader: Please be aware that this document only covers the specific use of that single user, mixes in quite a bit of unrelated stuff, and misses some quite important information. Prefer using the official documentation instead as it covers important details + stays up to date.

Consider this my last post on this topic and expect future answers on other questions to be much shorter.

**Reply by ksaadDE** [reply]
Hello Sir,

please refrain from spreading false information publicly. This is your last legal warning. Further steps ahead in that direction will be met with full force of law. 

Again, I politely reject your contract agreement offer, like I said the project is a tiny, private project. Seemingly your greed does not fulfill the ethics for my contractor agreements. 

Have a good week. 

Kind Regards
